### ISSUE TITLE
Avoid creation of NoopMeters

### ISSUE THREAD

**Issue by lenin-jaganathan** [micrometer-metrics/micrometer#4634]
**Please describe the feature request.**
Currently, when the register method is called it invokes the `registerMeterIfNecessary` and then `getOrCreateMeter`. [getOrCreateMeter](https://github.com/micrometer-metrics/micrometer/blob/061dbf69e908ce530e11b6e34082b01197263df6/micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java#L611C19-L611C35) applies the meter-filters and if any of the filters reply with a `MeterFilterReply.DENY` then a NoopMeter is created and returned. For every `MeterFilterReply.DENY` we go on creating a new NoopMeter which seems unnecessary. Instead, we should have DefaultInstances created for each of the NoopMeter and the same instance should be returned. 

**Rationale**
NoopMeter as the name suggests no-op and does nothing. So returning a default instance should be fine.

**Comment by shakuzen** [comment]
No-op meters still have a unique ID, so I'm not sure I see what optimization would make sense here. Can you elaborate?

**Comment by lenin-jaganathan** [comment]
Yeah, I have seen it. Do you know why we have Unique ID's in Noop Meters? I guess they are never added to the registries, so not sure where this ID will be useful.

**Comment by jonatan-ivanov** [comment]
I think it is used next time when you try to register the same meter that is already denied by the filter?

**Comment by lenin-jaganathan** [comment]
That's exactly the reason why I have created this issue.

When we see the meter is not accepted, [we apply the noopBuilder function and immediately return this](https://github.com/micrometer-metrics/micrometer/blob/061dbf69e908ce530e11b6e34082b01197263df6/micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java#L626). So, the next when someone tries to create the same meter, it goes through all the meter filter chain and returns a new Noop Instance instead.

**Comment by shakuzen** [comment]
I see now. We're not adding the NoopMeters to the meterMap, so the previously created instance isn't ever returned from getOrCreateMeter. I don't remember if that was intentional or not. It could be reasonable to change that.

**Comment by lenin-jaganathan** [comment]
Mostly useful when dealing dependant libs creating meters and filters deny them. Typically happens in spring (or similar environments) where we want to disable some of the redundant metrics. 

It also helps in optimizing meters with dynamic tags as the same of instance of NoopMeter can be used as reference.

**Comment by shakuzen** [comment]
> We're not adding the NoopMeters to the meterMap, so the previously created instance isn't ever returned from getOrCreateMeter.

I thought this might be as simple as adding the NoopMeters to the meterMap, and indeed that resolves this issue, but it has a significant side effect. MeterRegistry#getMeters (used in publishing typically) will return NoopMeters then, but currently it doesn't. We could start maintaining a separate noopMeterMap and check both meterMap and that in getOrCreateMeter, perhaps.

**Comment by lenin-jaganathan** [comment]
> I thought this might be as simple as adding the NoopMeters to the meterMap.
This would complicate things unnecessarily.

> We could start maintaining a separate noopMeterMap
This would again add to the problem that earlier the NoopMeters were at least Garbage collectible and by adding them to map we are keeping reference and might cause OOM.

**Comment by shakuzen** [comment]
```java
registry.config().meterFilter(MeterFilter.denyNameStartsWith("ignored"));
Counter a = registry.counter("ignored.one", "tag", "value");
Counter b = registry.counter(ignored.two");
```

You seem to want to return a Meter with a shared Id for all NoopMeters of the same type (e.g. Timer, Counter, etc.). Concretely, if I'm understanding what you want, given the above code, you want `a` and `b` to be the same value which is an instance of a NoopCounter returned for all NoopCounters regardless of name and tags. I can see how for typical uses, this won't be a problem in practice, but it would be breaking for getting the name of a returned meter or its tags to stop working. I don't know when or why users might be doing that with potentially no-op meters, but it seems like a valid thing to expect to be able to do given the current API. Right now you can rely on the returned meter having the right Id for what you registered regardless of if it is a no-op or not. I'm not sure how easily we could change that. Hence I was looking for solutions that maintained the same guarantees we have now while avoiding the redundant creation of no-op meters every time getOrCreate was called.

So, we would need to consider use cases that would be broken by a change as described above. We would need to weigh that against the benefit. Note that the `Id` is going to be instantiated regardless.
